### PR TITLE
Added Blum Goldwasser cryptosystem

### PR DESCRIPTION
**Pull request by czgdp1807**
#### References to other Issues or PRs
https://en.wikipedia.org/wiki/Blum%E2%80%93Goldwasser_cryptosystem

#### Brief description of what is fixed or changed
bg_private_key, bg_public_key, encipher_bg have
been added to sympy.crypto.crypto. These methods
follow Blum–Goldwasser cryptosystem.

#### Other comments
Documentation and tests will be added soon

#### Release Notes

* crypto
  * Added Blum Goldwasser cryptosystem

### CODE REVIEW

File: sympy/crypto/crypto.py [hunk 1]
```diff
@@ -28,7 +28,7 @@
 from sympy.polys.polytools import gcd, Poly
 from sympy.utilities.misc import filldedent, translate
 from sympy.utilities.iterables import uniq
-from sympy.utilities.randtest import _randrange
+from sympy.utilities.randtest import _randrange, _randint
```
Comment (asmeurer) [inline sympy/crypto/crypto.py]: These functions are supposed to be used for the test suite. This is the same as just using `random.Random`.

**Comment by czgdp1807** [comment]
So should I directly use the `random.Random`? I saw that in GM cryptosystem `_randrange` was used.

File: sympy/crypto/crypto.py (hunk 1 above)
Comment (czgdp1807) [reply sympy/crypto/crypto.py]: So should I directly use the random.Random? I saw that in GM cryptosystem _randrange was used.

**Comment by czgdp1807** [comment]
@asmeurer Please review the changes I have made recently.

File: sympy/crypto/crypto.py [hunk 2]
```diff
@@ -2246,3 +2246,199 @@ def decipher_gm(message, key):
...
+################ Blum–Goldwasser cryptosystem  #########################
+
+def _pow_mod(a, b, m):
+    """
+    Calculates (a**b) % m in an efficient manner.
+
+    Parameters
+    ==========
+
+    a: the base, a positive integer
```
Comment (oscargus) [inline sympy/crypto/crypto.py]: Please see the examples. Should be more like
```
a : positive integer
    The base
```

File: sympy/crypto/crypto.py [hunk 3]
```diff
@@ -2246,3 +2246,199 @@ def decipher_gm(message, key):
...
+
+def bg_public_key(p, q):
+    """
+    Calculates public keys from private keys.
+
+    The function first checks the validity of
+    private keys passed as arguments and
+    then returns their product.
+
+    Parametes
```
Comment (oscargus) [inline sympy/crypto/crypto.py]: Parameters. Missing r.

Comment (oscargus) [reply sympy/crypto/crypto.py]: Missing an "r" in Paramete**r**s. :-)

Comment (czgdp1807) [reply sympy/crypto/crypto.py]: Oh yes, I missed that.

File: sympy/crypto/crypto.py [hunk 4]
```diff
@@ -2246,3 +2246,202 @@ def decipher_gm(message, key):
         m <<= 1
         m += not b
     return m
+
+################ Blum–Goldwasser cryptosystem  #########################
+
+def _pow_mod(a, b, m):
```
Comment (jksuom) [inline sympy/crypto/crypto.py]: How about using Python's [builtin pow](https://docs.python.org/3.7/library/functions.html#pow)?

Comment (czgdp1807) [reply sympy/crypto/crypto.py]: Thanks for this. I am going to update.

**Comment by oscarbenjamin** [comment]
This looks good to me. There is an unresolved comment from @asmeurer but it looks like it can be merged anyway.

I'll merge soon if no one objects.

File: sympy/crypto/crypto.py [hunk 5]
```diff
@@ -2246,3 +2246,174 @@ def decipher_gm(message, key):
...
+    x = r**2 % key
+    x_L = pow(int(x), int(2**L), int(key))
+
+    rand_bits = []
+    for k in range(L):
+        rand_bits.append(x % 2)
+        x = x**2 % key
+
+    encrypt_msg = []
+    for (m, b) in zip(enc_msg, rand_bits):
```
Comment (jksuom) [inline sympy/crypto/crypto.py]: Would it be possible to use something like `encrypt_msg = [m^b for m, b in zip(enc_msg, rand_bits)]`?

**Comment by oscarbenjamin** [comment]
@czgdp1807 if you respond to @jksuom's comment I'll threaten to merge again :)

File: sympy/crypto/crypto.py (hunk 5 above)
Comment (czgdp1807) [reply sympy/crypto/crypto.py]: @jksuom My apologies for late reply. I think both ways are correct and I think yours is better. @oscarbenjamin If you also agree then I will make the change.

Comment (oscarbenjamin) [reply sympy/crypto/crypto.py]: I agree with @jksuom

**Comment by oscarbenjamin** [comment]
To be clear I don't know the crypto module in SymPy and I don't know the Blum Goldwasser cryptosystem. This just looks like a good patch to me and it would seem a shame for it to be abandoned. I have recently been through and labelled all 600 open PRs by topic and I saw many that looked like good patches that someone wrote that were just forgotten later. I think it's just because reviewing patches takes time and sometimes contributors are sending patches for areas that no regular contributors are focussed on.

My main concern about adopting a patch from an area I don't know that well is whether or not it represents a reasonable thing to include in SymPy at all. Three others have reviewed this patch suggesting minor improvements and no one has objected to the general concept so I will assume that this does represent a reasonable inclusion once the patch is good enough.

To me the patch looks good enough. However @jksuom has made a suggestion for improvement and it is a reasonable suggestion so if you do that and no one objects I will merge this.

**Comment by czgdp1807** [comment]
@oscarbenjamin Yes, I will surely update the PR. Actually, @jksuom suggestion is a good one to simplify the code. However, I thought to confirm since, any changes made will start another set of tests, which may fail due to one reason or other. Hence, confirming from two members sounded good to me. Since, we all agree I will make the changes as soon as possible.

**Comment by czgdp1807** [comment]
@oscarbenjamin @jksuom I have updated the PR. Please provide reviews, if any?

**Comment by oscarbenjamin** [comment]
Thanks for this. I'm merging it.

**Pull request merged**